### ISSUE TITLE
the problem of loss can not converge

### ISSUE THREAD

**Issue by ckqsars** [balancap/SSD-Tensorflow#279]
I found the reason why the loss can not converge.
In the  loss function you div the batch_size. 
however,
We should div the number of the positive labels and negative labels. 
Then  the loss will converge

the code change like:
 with tf.name_scope('cross_entropy_pos'):
            loss = tf.nn.sparse_softmax_cross_entropy_with_logits(logits=logits,
                                                                  labels=gclasses)
            loss = tf.div(tf.reduce_sum(loss * fpmask), fn_neg, name='value')
            tf.losses.add_loss(loss)

the fn_neg in the negative loss is the number of negative.

**Comment by HongyiDuanmu26** [comment]
In the original paper, loss should be divided by N( the number of matched default boxes), which means N is the number of positive masks.
Also, though, it is definitely difference between paper and codes, do you get a good result or find it easily to converge after changing this?
Thanks a lot. I have been struggling for this problem for days.

**Comment by ckqsars** [comment]
@HongyiDuanmu26 You can use the loss in the paper and it will easily to converage.

**Comment by birham-red-bd** [comment]
@ckqsars Did you solve the problem that the loss does not converge? Can you give me some help? I tried lots of methods, but the loss is still very high like 22.2, and the loss did not decrease even though I trained the model for the long time.

**Comment by ckqsars** [comment]
@songhuizhong you can see the change in the first comment.

**Comment by birham-red-bd** [comment]
@ckqsars yeah, I saw that, but the code you provided is kind of confusing, could you explain it in more detail? Thanks

**Comment by ZhenyF** [comment]
@ckqsars  Is your "fn_neg" mean "n_positives" in the code?

**Comment by ckqsars** [comment]
@ZhenyF   I think the fn_neg in the  code is the number of  negative samples.

**Comment by Just-A-Peek** [comment]
@ckqsars could you please tell me your mAP after loss function has been changed.

**Comment by fanbinqi** [comment]
@ckqsars thanks a lot,I have a question that I change the code like you said,in this code, the number of negative samples is n_neg,but i got some error,and cant train the network?could you give us your loss code?

**Comment by xy524679663** [comment]
@ckqsars  thanks a lot! I changed my code as u said and got great results on my own dataset

**Comment by jianyuheng** [comment]
Hi! how did you solve your problem? i was puzzled about "n_neg".

**Comment by liruixl** [comment]
Emmmm，mine does not converge，the loss stays about 30. 
Is there any other easier code about SSD，everyone? I‘m stupid...

**Comment by fanbinqi** [comment]
> Emmmm，mine does not converge，the loss stays about 30.
> Is there any other easier code about SSD，everyone? I‘m stupid...

ssd-pytorch,

**Comment by liruixl** [comment]
> > Emmmm，mine does not converge，the loss stays about 30.
> > Is there any other easier code about SSD，everyone? I‘m stupid...
> 
> ssd-pytorch,

thanks

**Comment by ckqsars** [comment]
@fanbinqi 
'        nmask = tf.logical_and(nmask, nvalues < max_hard_pred)
        fnmask = tf.cast(nmask, dtype)

        # Add cross-entropy loss.
        with tf.name_scope('cross_entropy_pos'):
            loss = tf.nn.sparse_softmax_cross_entropy_with_logits(logits=logits,
                                                                  labels=gclasses)
            loss = tf.div(tf.reduce_sum(loss * fpmask), batch_size, name='value')
            tf.losses.add_loss(loss)

        with tf.name_scope('cross_entropy_neg'):
            loss = tf.nn.sparse_softmax_cross_entropy_with_logits(logits=logits,
                                                                  labels=no_classes)
            loss = tf.div(tf.reduce_sum(loss * fnmask), batch_size, name='value')
            tf.losses.add_loss(loss)'
this is primary loss code

**Comment by ckqsars** [comment]
`        fn_neg = tf.cast(n_neg,dtype)
        fn_positives = tf.cast(n_positives, dtype)

        # Add cross-entropy loss.
        with tf.name_scope('cross_entropy_pos'):
            loss = tf.nn.sparse_softmax_cross_entropy_with_logits(logits=logits,
                                                                  labels=gclasses)
            loss = tf.div(tf.reduce_sum(loss * fpmask), fn_positives, name='value')
            tf.losses.add_loss(loss)

        with tf.name_scope('cross_entropy_neg'):
            loss = tf.nn.sparse_softmax_cross_entropy_with_logits(logits=logits,
                                                                  labels=no_classes)
            loss = tf.div(tf.reduce_sum(loss * fnmask), fn_neg, name='value')
            tf.losses.add_loss(loss)`

and this is my code of loss

**Comment by nizhanzhu** [comment]
@ckqsars  How many train steps the loss can converge after your code change about loss? thaknk you !

**Comment by alwaysPKU** [comment]
> ` fn_neg = tf.cast(n_neg,dtype)
> fn_positives = tf.cast(n_positives, dtype)
> 
> ```
>     # Add cross-entropy loss.
>     with tf.name_scope('cross_entropy_pos'):
>         loss = tf.nn.sparse_softmax_cross_entropy_with_logits(logits=logits,
>                                                               labels=gclasses)
>         loss = tf.div(tf.reduce_sum(loss * fpmask), fn_positives, name='value')
>         tf.losses.add_loss(loss)
> 
>     with tf.name_scope('cross_entropy_neg'):
>         loss = tf.nn.sparse_softmax_cross_entropy_with_logits(logits=logits,
>                                                               labels=no_classes)
>         loss = tf.div(tf.reduce_sum(loss * fnmask), fn_neg, name='value')
>         tf.losses.add_loss(loss)`
> ```
> and this is my code of loss

why without the localization loss?

**Comment by navy63** [comment]
Thank you very much. It's very reasonable to understand what you said. I'll run it in a few days to see if there's any improvement in mPAP.

**Comment by TXYUU** [comment]
I have a problem saying 'Tensor conversion requested dtype float32 for Tensor with dtype int32:'
loss_neg = tf.div(tf.reduce_sum(loss * fnmask), n_neg,name='value') Is n_neg with int32 different from tf.reduce_sum(loss * fnmask) with float32?

**Comment by ckqsars** [comment]
> I have a problem saying 'Tensor conversion requested dtype float32 for Tensor with dtype int32:'
> loss_neg = tf.div(tf.reduce_sum(loss * fnmask), n_neg,name='value') Is n_neg with int32 different from tf.reduce_sum(loss * fnmask) with float32?

You can try covert int32 to float32, Because I have not meet the problem,  maybe you can ask someone else.

**Comment by pntehan** [comment]
@ckqsars  Hi, I use your proposed loss function, but my loss becomes Nan, Although I decrease my learning rate, this case did not change. And I am sure my data is correct, Cuz I can train it on Faster RCNN.

**Comment by ckqsars** [comment]
> @ckqsars Hi, I use your proposed loss function, but my loss becomes Nan, Although I decrease my learning rate, this case did not change. And I am sure my data is correct, Cuz I can train it on Faster RCNN.
Maybe you can test on one or two pic to test the all training pipline.  I have try the loss function of the project and myself, the both dosen't become Nan.
